Bug fix to address work order receipt flow in work order sync mode execution

Ram-srini opened the pull request:
To fix issue - https://github.com/hyperledger/avalon/issues/412

Signed-off-by: Ramakrishna Srinivasamurthy <ramakrishna.srinivasamurthy@intel.com>

Ram-srini committed:
Bug fix to receipt flow in work order sync mode execution.
Address the git issue - https://github.com/hyperledger/avalon/issues/412

Signed-off-by: Ramakrishna Srinivasamurthy <ramakrishna.srinivasamurthy@intel.com>

manojgop commented:
@Ram-srini Could you update the commit message with details of the bug in work order receipt flow in sync mode.

manojgop commented on enclave_manager/avalon_enclave_manager/singleton/enclave_manager.py:
```diff
@@ -299,51 +299,52 @@ def __update_receipt(self, kv_helper, wo_id, wo_json_resp):
+        if receipt_entry is None:
+            receipt_entry = {
+                "params": {
+                    "receiptUpdates": []
```
Do we have any param like "receiptUpdates" in the spec ?

Ram-srini replied:
No. This is implementation specific, need to store the receipt updates in that tag.

manojgop commented on listener/avalon_listener/tcs_workorder_receipt_handler.py:
```diff
@@ -82,12 +82,22 @@ def WorkOrderReceiptCreate(self, **params):
             wo_receipt = self.kv_helper.get("wo-receipts", wo_id)
-            if wo_receipt is None:
+            wo_receipt = json.loads(wo_receipt)
+            if wo_receipt and "workOrderId" not in wo_receipt["params"]:
```
Under what scenario will workOrderId be not part of wo_receipt ? My understanding is receipt is for a specific work order. So a receipt does not exists without a work order

Ram-srini replied:
WorkOrderId is exists if work order receipt is created. Refer the git issues for description  https://github.com/hyperledger/avalon/issues/412

rranjan3 commented on enclave_manager/avalon_enclave_manager/singleton/enclave_manager.py:
```diff
@@ -299,51 +299,52 @@ def __update_receipt(self, kv_helper, wo_id, wo_json_resp):
-                updated_receipt))
-            logger.info("Receipt for the workorder id %s is updated to %s",
-                        wo_id, wo_receipt)
+        # If receipt is not created yet, add tag "receiptUpdates" to
```
Minor typo in function comment @292 wo_json_ >>> json response
@298 - I do not follow the description there

rranjan3 commented on enclave_manager/avalon_enclave_manager/singleton/enclave_manager.py:
```diff
@@ -299,51 +299,52 @@ def __update_receipt(self, kv_helper, wo_id, wo_json_resp):
+            self.private_key
+        )
+
+        # If it is first update to receipt
```
it is first update >>> is not the first update

rranjan3 commented on enclave_manager/avalon_enclave_manager/singleton/enclave_manager.py:
```diff
@@ -299,51 +299,52 @@ def __update_receipt(self, kv_helper, wo_id, wo_json_resp):
+                    "Receipt for the workorder id %s is completed " +
+                    "and no further updates are allowed",
+                    wo_id)
+                return
```
Can there be contradictory receipt updates coupled with COMPLETED ? Most probably no. Otherwise we might still want to check whats the latest one and not return.

Ram-srini replied:
always check the latest status of receipt, if it is COMPLETED then there won't be further updates to it.

rranjan3 reviewed (commented):
Please add some link to defect filed or some description otherwise to keep record of what was breaking and how it was fxed in the commit message. Also try compressing the commit title.

Ram-srini committed:
Fix stand alone mode test failure
Since listener config is hard-coded with container service name
and it is failing in standlone mode.
Added command line option to listener to read zmq_url and defact value
in config file is localhost and command line argument takes precedence.

Signed-off-by: Ramakrishna Srinivasamurthy <ramakrishna.srinivasamurthy@intel.com>
